odoo/addons/base/models/res_device.py

```diff
@@ -169,65 +129,163 @@ def __update_revoked(self):
-                    WHERE
-                        D2.user_id = D.user_id
-                        AND D2.session_identifier = D.session_identifier
-                        AND D2.platform IS NOT DISTINCT FROM D.platform
-                        AND D2.browser IS NOT DISTINCT FROM D.browser
-                        AND (
-                            D2.last_activity > D.last_activity
-                            OR (D2.last_activity = D.last_activity AND D2.id > D.id)
-                        )
-                        AND D2.revoked IS NOT TRUE
-                )
-                AND D.revoked IS NOT TRUE
+                L.revoked IS NOT TRUE
+            GROUP BY
+                session_identifier,
+                ip_address,
+                user_agent
         """
 
+    def init(self):
+        tools.drop_view_if_exists(self.env.cr, self._table)
+        self.env.cr.execute('CREATE OR REPLACE VIEW %s AS (%s)' % (self._table, self._query))
+
+
+class ResSession(models.Model):
+    _name = 'res.session'
+    _inherit = ['res.device']
+    _description = 'Session'
+    _auto = False
+    _order = 'last_activity desc'
+
+    device_ids = fields.One2many('res.device', compute='_compute_session_info', readonly=True)
+    current_device_id = fields.One2many('res.device', compute='_compute_session_info', readonly=True)
+    ip_address = fields.Char('IP Address', compute='_compute_session_info', readonly=True)
+    user_agent = fields.Char('User Agent', compute='_compute_session_info', readonly=True)
+
+    is_current = fields.Boolean('Current', compute='_compute_is_current', readonly=True)
+
+    def _compute_is_current(self):
+        if not request:
+            self.is_current = False
+            return
+        session_identifier = request.session.sid[:STORED_SESSION_BYTES]
+        for session in self:
+            session.is_current = session.session_identifier == session_identifier
+
+    def _compute_session_info(self):
+        session_map = {}
+        for session_identifier, device_line_ids in self.env['res.device']._read_group(
+            domain=[('session_identifier', 'in', self.mapped('session_identifier'))],
+            groupby=['session_identifier'],
+            aggregates=['id:recordset'],
+        ):
+            session_map[session_identifier] = device_line_ids.sorted('last_activity', reverse=True)
+
+        for session in self:
+            device_ids = session_map[session.session_identifier]
+            session.device_ids = device_ids
+            if session.is_current:
+                # The user will see the information for the device he is
+                # currently using, even if there are more recent logs for
+                # another device.
+                current_device_id = device_ids.filtered('is_current')
+            else:
+                current_device_id = device_ids[0]  # because order is `last_activity desc`
+            session.ip_address = current_device_id.ip_address
```

Comment:
You forgot to assign `current_device_id`
```suggestion
            session.current_device_id = current_device_id
            session.ip_address = current_device_id.ip_address
```